### ISSUE TITLE
[feature request] Search field for how-to page

### ISSUE THREAD

**Issue by darigovresearch** [ONEARMY/community-platform#1088]
**Is your feature request related to a problem? Please describe.**
A user in the discord mentioned that now there are many how-tos it would be helpful to have a search field to help find relevant how-tos a user is looking for, particularly when it is an old how-to

**Describe the solution you'd like**
A search field at the top of the page that would narrow down the visible how-tos from the database

**Describe alternatives you've considered**
Can use ctrl+f in the browser but does not always work as not all how-tos are loaded onto the page

**Additional context**
n/a

**Comment by chrismclarke** [comment]
I really like this suggestion. There's a few different ways we could go about this:

1. Manual client-side search function that tries to match text from the input box against a few specific fields, such as title, description, and maybe howto steps. Pros: very quick to implement. Cons: search basic and harder to use elsewhere

2. Server-side search via algolia or elasticsearch. We use algolia in the platform already for map location lookup, so could use the existing account to also try add search against our own content. Pros: efficient search. Cons: not sure how well pricing scales

3. Client-side search via lib. I'd be leaning towards [fuse.js](https://fusejs.io/) for this, although there are plenty of [other options](https://www.npmtrends.com/fuse.js-vs-fuzzysearch-vs-lunr-vs-search-index-vs-elasticlunr-vs-js-search) also. Pros: efficient search. Cons: wouldn't work as well if we had very large datasets.

Personally I'd be leaning towards option (3), but open if anyone wants to take this on and explore any other options

**Comment by danitrod** [comment]
I'd be willing to work on this, I think we can start with option 3, then eventually switch to server-side search if need be later on. 

About the layout of the input field, I think it could be similar to the filter by tags button, maybe with a magnifier icon aside, and could be positioned in 1. or 2. (before the filter button) like the image below. Any thoughts?

<img width="1437" alt="Captura de Tela 2021-02-04 às 11 12 08" src="https://user-images.githubusercontent.com/45438149/106905850-597a7280-66db-11eb-8175-83ff2f23094d.png">

**Comment by chrismclarke** [comment]
Sounds great, yes I think for the howto's as there's only 100 or so for now and unlikely to grow too big too soon I think fuse.js will be a solid option for some time. We have a pretty similar layout already for events so I'd suggest starting with that and then if design team want some changes we can update.

![image](https://user-images.githubusercontent.com/10515065/106943204-fa682e00-66d9-11eb-8dbe-5d28eea686bb.png)

I think the component used for the event search is pretty hard-wired for use with algolia so you will probably be best off creating a new search component and if we later decide we want to combine just refactor then (although to me it seems they're trying to do sufficiently different things to justify multiple components. Then up to you if you want the search box to just handle the input and pass back the results to filter in the howto store, or if we have something more like a `fuseSearch` component that takes the list of data to search as a prop and performs the search function within the component itself, passing back the results.

**Comment by darigovresearch** [comment]
If possible it would be great to have that when you searched something in the search box it would generate a unique URL for the results so that a search results can easily be shared

Would this be easy to incidentally add when building the search feature?

**Comment by chrismclarke** [comment]
@darigovresearch @davehakkens 
Thanks to the work by @danitrod this feature has just gone into preview, you can check it out here:
https://onearmy-next--pr1098-1088-howto-search-96tkaftb.web.app/how-to

I can also see from the PR that there's some nice flexibility for which content we search for and how it's weighted, so feel free to share your thoughts if you have any regarding the specific implementation. Here's what is currently specified:

```
const HOWTO_SEARCH_WEIGHTS = [
  { name: 'title', weight: 0.5 },
  { name: 'description', weight: 0.3 },
  { name: 'steps.title', weight: 0.15 },
  { name: 'steps.text', weight: 0.05 },
]
```
It also has support for 'fuzzy' matching which I think is pretty cool (just tried searching for bicicle), and can be further refined.

I really like the idea of also saving the result within a URL as suggested, either for sharing a link or for perhaps when you navigate to the howto it can also highlight the section that matched the search criteria. Perhaps we can open another issue once we have this merged in.

**Comment by davehakkens** [comment]
looks neat, and works smooth! Thanks @danitrod  
Also loving this preview thing @chrismclarke 

Ready to merge besides one small request, while we are at it. (only if it's easy otherwise for a future issue)
Is there a way to include the 'creator' in the search?
A few people requested a few to filter content from a specific creator, this could be a good way to add that

**Comment by danitrod** [comment]
For sure @davehakkens; I'll update the search weights to consider the creator before the merge. What do you think of the following weights for the search?
```
0.5 - Title
0.2 - Description
0.15 - Creator
0.1 - Steps titles
0.05 - Steps descriptions
```

About the URL updating, I think it's best to separate into a new issue, but I'll also be happy to work on that.

**Comment by davehakkens** [comment]
Sounds good to me!

**Comment by chrismclarke** [comment]
Adds to 1 so that's definitely good enough for me too :D
I'll merge the current changes (can see the weights have already been updated) and open a new issue for the search params

**Comment by chrismclarke** [comment]
PR merged and new issue opened for search URL #1100
